Implement SSL

argysamo opened the pull request:
Introduce backwards compatible SSL.
- [script.py] Add optional argument for ssl-private_key, ssl-public_key and ssl-ca
- [web.py] Add optional ssl_context
- [conftest.py] Pytest utilities to test ssl
- [script_test.py/web_test.py] Add ssl testing

argysamo committed:
Implement SSL

Introduce backwards compatible SSL.
- [script.py] Add optional argument for ssl-private_key, ssl-public_key and ssl-ca
- [web.py] Add optional ssl_context
- [conftest.py] Pytest utilities to test ssl
- [script_test.py/web_test.py] Add ssl testing

albertodonato commented on prometheus_aioexporter/script.py:
```diff
@@ -164,6 +187,20 @@ def _configure_registry(self, include_process_stats: bool = False) -> None:
             )
 
+    @staticmethod
+    def _configure_ssl(args: argparse.Namespace) -> Optional[SSLContext]:
```
I'd rename this to something like `_get_ssl_context`, and drop the `@staticmethod`, so it's possible to just call it on `self` below.

Also, you can use `| None` rather then `Optional` (here and other places)`

argysamo replied:
thanks - fixed!

albertodonato commented on tests/conftest.py:
```diff
@@ -0,0 +1,48 @@
+def ssl_context(tls_certificate):
+    ssl_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
+    tls_certificate.configure_cert(ssl_ctx)
+    return ssl_ctx
```
please use `yield` here and below

argysamo replied:
thanks - fixed!

albertodonato commented on tests/script_test.py:
```diff
@@ -73,42 +79,114 @@ def test_setup_logging(self, mocker):
+        assert isinstance(exporter.ssl_context, SSLContext)
+        assert len(exporter.ssl_context.get_ca_certs()) == 1
+
+    @pytest.mark.xfail
```
Why is this expected to fail?

argysamo replied:
Now I remember why I've added (and forgot about this). Even when I run unit tests from `main` branch (i.e. without any changes), that specific unit test fails. It's a mystery that this unit tests succeeds when run on Github.
I've removed the `xfail` and tests succeed now. I am resolving the conversation.

albertodonato commented on prometheus_aioexporter/script.py:
```diff
@@ -134,6 +139,24 @@ def get_parser(self) -> argparse.ArgumentParser:
+            type=str,
+            dest="ssl_ca",
+            help="full path to the ssl certificate authority (CA)",
+        )
```
You can make those `type=argparse.FileType()`, so that file existence is checked. Also `dest=` can be dropped as it defaults to the same values.

argysamo replied:
Thanks so much. I am using the `argparse.FileType()` now. However, unistests started failing locally now. `trustme` generates certificates in a temporary folder. The `load_verify_locations` of the `ssl` module fails because of that temporary file. I will continue testing.

argysamo replied:
I found and fixed the issue! 
The `load_verify_locations` of the `ssl` module expects the path to the certificate components as `str`. I was providing it with `FileType` objects and this is why it was failing. It was related with the temporary file and `trustme`. I am resolving this converasation

albertodonato reviewed (commented):
Thanks for the contribution! 

Changes look good, I have a few minor comments.

albertodonato commented on tests/conftest.py:
```diff
@@ -0,0 +1,48 @@
+import ssl
+
+import pytest
+import trustme
```
this should be added to test dependencies

argysamo replied:
oops - added!

argysamo committed:
Fix PR comments

- [conftest.py]: Replace all `return` with `yield` statements
- [pyproject.toml]: Add `trustme` as test dependency
- [script.py]: `_configure_ssl` is now `_get_ssl_context` (not a staticmethod anymore). SSL arguments are of type `FileType` (not `str`).
- [script_test.py]: Remove the `xfail` decorator
- [web.py and script.py]: Replace `Optional` with `| None`

argysamo committed:
Reformat files with black

argysamo committed:
Fix FileType issue

- args.FileType returns a FileType object. `load_cert_chain` expects public/private key paths as `str`. `name` attribute (of `FileType`) returns exactly what is needed.

argysamo committed:
Fix lint issues

- Sort imports

albertodonato commented on prometheus_aioexporter/script.py:
```diff
@@ -3,6 +3,8 @@
 from collections.abc import Iterable
 import logging
+import ssl
+from ssl import SSLContext
```
Given yoou're importing `ssl`, you can just use `ssl.SSLContext` below and drop this extra import

argysamo replied:
nice catch - removed!

albertodonato commented on prometheus_aioexporter/script.py:
```diff
@@ -164,6 +181,21 @@ def _configure_registry(self, include_process_stats: bool = False) -> None:
+            return None
+        cafile = None
+        if args.ssl_ca:
+            cafile = args.ssl_ca.name
+        ssl_context = ssl.create_default_context(
+            purpose=ssl.Purpose.CLIENT_AUTH, cafile=cafile
+        )
+        ssl_context.load_cert_chain(
+            args.ssl_public_key.name, args.ssl_private_key.name
```
since  `FileType` returns an open file, it'd be good to `close()` them here

argysamo replied:
nice one! I am closing all of them in place.

albertodonato reviewed (commented):
Thanks for the fixes! A couple of final (and minor) comments

albertodonato commented:
FWIW the lint issue should go away if you rebase on main

argysamo committed:
PR review comments and README.rst

- [script.py]: Remove `from ssl import SSLContext` statement. Close ssl related FileType objects.
- [README.rst]: Add references to the SSL arguments

argysamo commented:
> FWIW the lint issue should go away if you rebase on main

~I have branched out main branch and main branch wasn't updated since. Yet, I tried to rebase, but there wasn't anything to rebase onto.~
Ignore my comment above. I wasn't fetching with `--all` flag and I was only fetching my origin remote. I have now rebased on the latest main!

albertodonato commented:
Thank you very much!

argysamo commented:
Hello Alberto,
Thank you so much for you kind words and your PR feedback. 
A question: Do you plan to create a new release of the latest changes? I’d love to use the ssl with your query-exporter.

Cheers,
Argyris

--
Argyrios Samourkasidis
Αργύριος Σαμουρκασίδης

> On 23 Oct 2023, at 22:37, Alberto Donato ***@***.***> wrote:
> 
> 
> Thank you very much!
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/albertodonato/prometheus-aioexporter/pull/6#issuecomment-1775980966>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABAHMKPAIQQVXQTANMFQQZ3YA3IQNAVCNFSM6AAAAAA6JPZSWSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZVHE4DAOJWGY>.
> You are receiving this because you authored the thread.
>

albertodonato commented:
I plan to make a 2.0 release soon which will also include the changes I made for `MetricConfig` (which are backwards-incompatible, hence the change). Will have to update query-exporter too for this change.

argysamo commented:
thanks so much. Please let me know if there's anything I could help with!
